Facet voxel traverse tool

zxkjack123 opened the pull request:
This PR adds some functions to perform voxel traverse for ray tracing. And then add a function to calculate the voxels intersect with given facet based on the ray tracing.

Based on [A Fast Voxel Traversal Algorithm for Ray Tracing](http://www.cse.yorku.ca/~amana/research/grid.pdf), a modified ray traverse method was used to calculate the voxels intersect with a given ray (line segment). Several rays are used to calculate the voxels that intersect with the facet. A series of rays are constructed based on both the facet and the mesh grid to make sure all the wanted voxels will be found. Combining all the found voxels for each ray, we get the voxels intersect with the facet. 

These voxels will be used to construct the bounding volume for source sampling. But currently these functions are separate from MOAB. Interface of MOAB with PyNE is still in need.

Major changes:

- Functions used to calculate the voxels that intersect with given ray.

- Functions used to calculate the voxels that intersect with given facet (triangle).

- Test functions for the function of ray/facet voxel traverse.

zxkjack123 committed:
add functions for facet voxel traverse

zxkjack123 committed:
add functions for facet voxel traverse

zxkjack123 committed:
add test function for ray traverse and facet traverse

zxkjack123 committed:
add test function for ray traverse and facet traverse

zxkjack123 committed:
add news file

scopatz commented on pyne/utils.py:
```diff
@@ -190,3 +191,534 @@ def remove(path):
+
+class Point(object):
+    """
+    Class point.
```
Can you please be more descriptive here.

scopatz commented on pyne/utils.py:
```diff
@@ -190,3 +191,534 @@ def remove(path):
+
+    def __str__(self):
+        """
+        Print point according to the coordinate.
```
This doesn't actually print

scopatz commented on pyne/utils.py:
```diff
@@ -190,3 +191,534 @@ def remove(path):
+    v_z = d_z / dist
+    return [v_x, v_y, v_z]
+
+def _find_next_grid_1D(cor, v_1D, bound):
```
PEP8: variable and function names should not have uppercase characters

scopatz commented:
Thanks for putting this in @zxkjack123 - other than a couple of minor issues, this looks good to me!

zxkjack123 committed:
some clean up

zxkjack123 committed:
some clean up

zxkjack123 commented:
@scopatz I have made some changes according to your comment.

scopatz commented on pyne/utils.py:
```diff
@@ -190,3 +191,535 @@ def remove(path):
+
+##################################
+### Facet voxel traverse tools ###
+##################################
```
hmm it is a little weird to me that these are all in `utils.py` rather than their own module, since this seems like such a self-contained piece of code.  Is there a reason for them living in `utils.py` that I am not seeing?

zxkjack123 replied:
These functions are indeed a self-contained piece of code. However, among all these functions, only one function, the `_facet_voxel_traverse`, will be called. 

I am not sure where should I put all these functions. I put them in `utils.py` becasue they just provide an useful function as a tool.

scopatz replied:
I think you should make a new module that has an appropriate name for this capability

zxkjack123 committed:
move voxel traversal algorithm functions to moudle vta

zxkjack123 committed:
move voxel traversal algorithm functions to moudle vta

zxkjack123 commented:
@scopatz I have moved the voxel traversal algorithm function to module `vta.py`.

scopatz commented on news/facet_voxel_traverse_tool.rst:
```diff
@@ -0,0 +1,17 @@
+**Added:**
+
+* Functions used to calculate the voxels that intersect with given ray.
```
Can you add that these are in the `vta` module please?

scopatz commented on pyne/vta.py:
```diff
@@ -0,0 +1,535 @@
+from __future__ import division
```
Please provide a module level docstring

scopatz commented on tests/test_vta.py:
```diff
@@ -0,0 +1,49 @@
+from numpy.testing import assert_array_equal, assert_array_almost_equal
+
+from pyne import vta
+import numpy as np
```
PEP8: numpy imports should come before pyne imports.  Please move this line up.

scopatz commented:
Thanks for updating! Can you also please add a documentation hook to the `docs/pyapi/` directory? It should be called `vta.rst`. You can just copy one of the other files in that directory over and then edit the contents to say vta instead.

zxkjack123 committed:
clean up and add docstring for vta module

zxkjack123 committed:
add vta.rst in docs/pyapi

katyhuff commented:
This seems like maybe it's ready to merge. Thoughts @zxkjack123 ?

zxkjack123 commented:
Thanks @katyhuff. I will let this PR to be discussed in a group meeting in next one or two weeks. I'd like to deal with this PR after the discussion in the group meeting.

pshriwise commented on pyne/vta.py:
```diff
@@ -0,0 +1,551 @@
+"""Purpose:
+
+This module provides a modified ray traverse method, which is modifed based on
+http://www.cse.yorku.ca/~amana/research/grid.pdf, to calculate the voxels
```
Some minor grammar here "to calculate which voxels intersect with a given ray (line segment)".

pshriwise commented on pyne/vta.py:
```diff
@@ -0,0 +1,551 @@
+Several rays are used to calculate the voxels that intersect with the facet.
+A series of rays are constructed based on both the facet and the mesh grid to
+make sure all the wanted voxels will be found. Combining all the found voxels
+for each ray, we get the voxels intersect with the facet.
```
"we get the voxels intersect with the facet" --> "we get the voxels intersecting with the facet"

pshriwise commented on pyne/vta.py:
```diff
@@ -0,0 +1,551 @@
+    Parameters:
+    -----------
+    bounds : list
+        Boundaries. List of the boundaries along x, y, and z directions.
```
This is a nested list, expected to be of the format 

`[[x_min, x_max],[y_min, y_max],[z_min,z_max]]`. 

I could see myself also expecting this to work with a nested list of the form 

`[[x_min, y_min, z_min],[x_max, y_max, z_max]]`

I think it's worth being specific about the expected structure of this input parameter in the documentation.

zxkjack123 replied:
I will add the expected structure here. And the expected structure will be:
```
        [[x_min, ..., x_max], [y_min, ..., y_max], [z_min, ..., z_max]]
 ```
The boundaries in each direction could be different (iints, jints and kints in mesh are different). Therefore, the format like:
```
[[x_min, z_min, z_min], ..., [x_max, y_max, z_max]]
``` 
will not work. 

Though we can construct a nested list that contains only the min and max boundaries of a mesh, it is not necessary.

pshriwise replied:
True. I was more suggesting that based on the documentation a user (or future developer) might not know how to structure the list if they wanted to use this function.

pshriwise commented on pyne/vta.py:
```diff
@@ -0,0 +1,551 @@
+        End point
+    v : list
+        Direction vector from start to end: [v_x, v_y, v_z].
+    bounds : list
```
Again, I think this could be a little more descriptive in terms of what is expected to be in this set of nested lists.

pshriwise commented on pyne/vta.py:
```diff
@@ -0,0 +1,551 @@
+        t_max_y = float('inf')
+    else:
+        t_max_y = (n_y_grid - tp.y) / v[1]
+    return [t_max_x, t_max_y]
```
The doc strings indicate that three directions will be returned, but only two are found here.

zxkjack123 replied:
Oops, I forgot to update the code here. It should return data in three directions. This will be fixed in next commit.

pshriwise commented on pyne/vta.py:
```diff
@@ -0,0 +1,551 @@
+        Modified z_idx.
+    tp : Point
+        Point after moving along the direction of v.
+    t_temp : float
```
Does this mean an updated length of the ray?

zxkjack123 replied:
Yes, it is an updated length of the ray. The distance between the start point and updated point position.

pshriwise replied:
It might be nice to provide some extra clarity about this. If I understand correctly, the returned `t_temp` value is the total distance traveled from the start of the ray, but that isn't used or provided as part of this function, which is a bit confusing.

I'll suggest something like "Updated travel length since the beginning of the ray." I'm not sure if that's better really 😞

Also, having different names for outgoing variables might be useful to provide clarity here.

pshriwise commented on pyne/vta.py:
```diff
@@ -0,0 +1,551 @@
+
+def _move_to_boundary(tp, t_min, v):
+    """
+    Move the point to the nearest boundary when Initialize the problem.
```
"Initialize" --> "initializing"

zxkjack123 replied:
It will be  fixed in the next commit.

pshriwise commented on pyne/vta.py:
```diff
@@ -0,0 +1,551 @@
+    """
+
+    # calculate the new coordinate
+    tp.x += t_min * v[0]
```
This only updates two dimensions?

zxkjack123 replied:
Sorry again. Another place I forgot to update. It should update three dimensions.

pshriwise commented on pyne/vta.py:
```diff
@@ -0,0 +1,551 @@
+                # point outside the mesh
+                return voxel_list
+            # add current voxel
+            voxel_list.append([x_idx, y_idx, z_idx])
```
It might be worth considering tuples here. You'll be looping over these values later (presumably) and won't want them to accidentally get changed.

pshriwise reviewed (changes requested):
This looks good @zxkjack123! It's a considerable amount of work and you've made it a very modular piece of code. It was very enjoyable to review 🏆 

I have a couple design comments a couple concerns/questions:

  - The `Point` class you've created comes in handy in so many places here! I think it would be worth considering the addition of other classes (such as `Direction` and `Voxel` classes) which will reduce the number of arguments you have to pass from function to function and make the code more clean. Additionally, it will make documentation easier! Instead of having to explain (perhaps repeatedly) that an argument is "a list containing a direction vector u,v,w" you can simply claim it was a `Direction` class object and refer them to the class documentation :)

  - There are some places I've pointed out where I think only 2 of 3 dimensions are being handled. I coule be missing something there but it seems odd and the unit tests are structured in such a way that if only the x & y dimensions are handled they might not catch those problems - so I'm wary of that

Overall, this looks very good. I'll take another look once some of these comments have been addressed.

zxkjack123 commented:
@pshriwise Thank you very much for your comments. You have pointed out that I forgot to update some places of the code, leaving the z direction unchanged. And the unit tests are designed only in x and y directions, therefore, these bugs are not detected by the tests. I will add a 3D unit test to this PR to make sure this PR gets a correct answer first. I will update this PR when I finish this step.

Then I will restructure the code, adding some classes such as direction and voxel to make the code cleaner and easier to read.

zxkjack123 committed:
add unit test in 3D

zxkjack123 committed:
add unit test in 3D

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,566 @@
+    for ray in rays:
+        temp_voxel_list = _ray_voxel_traverse(bounds, ray[0], ray[1])
+        # merge two lists
+        voxel_list = _merge_voxel_list(voxel_list, temp_voxel_list)
```
You can use a `set` and then it will take care of duplicates automatically

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,566 @@
+    c = _distance(A, B)
+    min_edge = min([a, b, c])
+    if a == min_edge:
+        # divide BC
```
```
if a == min_edge:
    target = A
    source = (B,C)
elif b == min_edge:
    target = B
    source = (A,C)
else
    target = C
    source = (A,B)
insert_points = _divide_edge(source, min_step)
rays = _create_rays_from_points(target, insert_points)
```

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,566 @@
+        List of points. These points are located on the line segment from start
+        to end.
+    """
+    dist = _distance(start, end)
```
```
num_segments = floor(dist/step) + 2
x_list = numpy.linspace(start.x, end.x,num_segments)
y_list = numpy.linspace(start.y,end.y,num_segments)
z_list - numpy.linspace(start.z,end.z,num_segments)
insert_points = []
for (x,y,z) in zip(x_list,y_list,z_list):
    insert_points.append(Point(x,y,z))
```

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,566 @@
+    min_step = float('inf')
+    for i in range(len(bounds)):
+        for j in range(len(bounds[i])-1):
+            step = bounds[i][j+1] - bounds[i][j]
```
add 3rd dimension

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,566 @@
+        The minimum grid step.
+    """
+    min_step = float('inf')
+    for i in range(len(bounds)):
```
check each dimension/direction independently because the grids are structured

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,566 @@
+    for i in range(len(bounds)):
+        for j in range(len(bounds[i])-1):
+            step = bounds[i][j+1] - bounds[i][j]
+            if min_step > step:
```
`min_step = min(min_step,step)`

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,566 @@
+    min_step: float
+        The minimum grid step.
+    """
+    min_step = float('inf')
```
```
for i in range(len(bounds)-1):
    min_step = min(min_step,bounds[i+1]-bounds[i])
``

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,566 @@
+        The minimum grid step.
+    """
+    min_step = float('inf')
+    for i in range(len(bounds)):
```
```
for bound_list in bounds:
   for bound_idx in range(len(bound_list)-1):
        min_step = min(min_step,bound_list[bound_idx+1]-bound_list[bound_idx])
```

gonuke replied:
maybe too clever:
```
for (upper,lower) in zip(bound_list[1:],bound_list[:-1]):
    min_step = min(min_step,upper-lower)
```

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,566 @@
+
+    Parameters:
+    -----------
+    bounds : list
```
elaborate on the structure of `bounds`

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,566 @@
+            # length are negtive, then there is no intersection, return. Ohterwise,
+            # choose the minimum length and move the point to that position.
+            # left, right, back, front, down, up
+            t_maxs = [0.0] * 4
```
```
for dir in [0,1,2]:
   if v[dir] == 0.0:
     t_maxs[dir*2] = float('inf')
     t_maxs[dir*2 +1]=float('inf')
   else:
      t_maxs[dir*2] = (bounds[dir][0] - tp.dir) / v[dir]       # tp.dir is not really defined with your current data structure
      t_maxs[dir*2+1] = (bounds[dir][-1] - tp.dir) / v[dir]
```

zxkjack123 commented on pyne/vta.py:
```diff
@@ -0,0 +1,566 @@
+##################################
+
+
+class Point(object):
```
Make Point iterable.

zxkjack123 replied:
Maybe just a numpy array of size 3.

pshriwise replied:
One option for this is to make a list of the x,y,z values and index that list. This would allow you to keep the nicely named attributes if you're referring to a single value, but still append to lists or combine points cleanly. 

```
def __iter__(self, key):
    l = list(self.x, self.y, self.z)
    return l[key]
```

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,566 @@
+    t_temp : float
+        Updated travel length. Distance from the start point to updated point.
+    """
+    t_min = min(t_maxs)
```
look to use `numpy.argmin`

gonuke replied:
```
update_idx = [0,0,0]
update_dir = nump.argmin(t_maxs)
update_idx[update_dir] = 1
[x_idx,y_idx,z_idx] = [x_idx,y_idx,z_idx] + update_idx
```

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,566 @@
+    n_x_grid = _find_next_grid_1d(tp.x, v[0], bounds[0])
+    n_y_grid = _find_next_grid_1d(tp.y, v[1], bounds[1])
+    n_z_grid = _find_next_grid_1d(tp.z, v[2], bounds[2])
+    if v[0] == 0.0:
```
see below for loop version of this

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,566 @@
+    n_grid = 0.0
+    flag = False
+    if v_1d > 0:
+        for i in range(len(bound)):
```
```
for bnd_pos in bound:
   if cor < bnd_pos:
       n_grid = bnd_pos
       flag  = True
       break
```

zxkjack123 commented on pyne/vta.py:
```diff
@@ -0,0 +1,566 @@
+    dist : float
+        The distance between the start point and the end point.
+    """
+    dist = math.sqrt(math.pow(end.x - start.x, 2) +
```
Use (end.x - start.x) * (end.x - start.x)

zxkjack123 committed:
change Point to a numpy array, change voxel id to tuple

zxkjack123 committed:
change Point to a numpy array, change voxel id to tuple

zxkjack123 committed:
change the voxel list to set

zxkjack123 committed:
change the voxel list to set

zxkjack123 committed:
change the calling of divide_edge

zxkjack123 committed:
change edge divide method

zxkjack123 committed:
change function to find minimum grid step

zxkjack123 committed:
change the t_maxs calculation, reduce code duplication

zxkjack123 committed:
put x_idx, y_idx, z_idx into idxs

zxkjack123 committed:
change a iteration variable name

zxkjack123 committed:
clean up

zxkjack123 commented:
Major changes:
- Use numpy array to replace the class `Point`.
- Put some variables, such as `x_idx,  y_idx, z_idx` into a list or numpy array. Make them iterable to reduce the code duplication.
- Change the method for rays creating. Divide the shortest triangle edge uniformly.

pshriwise commented on pyne/vta.py:
```diff
@@ -0,0 +1,481 @@
+    # idx could be -1, that means point outside the mesh
+    return idx
+
+def _cal_dir_v(start, end):
```
Could we afford to be a little more verbose here? Even `_cal_dir_v` --> `_calc_vec_dir` provides a more clarity.

pshriwise commented on pyne/vta.py:
```diff
@@ -0,0 +1,481 @@
+    -----------
+    cor : float
+        Coordinate of current point in the direction.
+    v_1d : float
```
Is there a particular reason this is a float rather than a Bool or Int?

zxkjack123 replied:
The `v_1d` here is a number in `[v_x, v_y, v_z]`, it is a float. It can be converted to a `bool` or `int` before the next `if` function. However, the conversion will not simplify the code. For the better understanding of the code, I keep this parameter to be a float.

pshriwise replied:
I think that's ok, yeah. The odd part to me was a float being used to make a binary choice in the code, but I see what you mean. It also comes in handy at the end of the function for setting the sign of INF. 

I'm going to propose something else to simplify the code here though. The for loops can be consolidated like so: 

```
if v_1d < 0: 
    bound = reversed(bound)
    cor = -cor

for bnd in bound:
    if cor < bnd_pos:
        n_grid = bnd_pos
        flag = True
        break
```

The for loop code can always be the same if the iterable is setup correctly.

pshriwise commented on pyne/vta.py:
```diff
@@ -0,0 +1,481 @@
+    tp : numpy array
+        An array of size 3. Temporary point position.
+    t_temp: float
+        Temporary travel lenght from the start.
```
lenght --> length

zxkjack123 replied:
Will be revised in next commit.

pshriwise commented on pyne/vta.py:
```diff
@@ -0,0 +1,481 @@
+    update_dir = list(t_maxs).index(t_min)
+    update_idx = [0, 0, 0]
+    update_idx[update_dir] = 1
+    idxs = [x + y for x, y in zip(idxs, update_idx)] 
```
Should this always be an addition?

zxkjack123 replied:
Found a bug here, the `update_idx[update_dir]` could be both `1` or `-1`. This is fixed in next commit.

pshriwise commented on pyne/vta.py:
```diff
@@ -0,0 +1,481 @@
+    Parameters:
+    -----------
+    tp : numpy array
+        An array of size 3. Temporary point. Actually, it is also the start point.
```
Not a huge deal but `temporary point` in a lot of the documentation is messing with me. In the context of these individual functions, the "temporary point" is the only point it knows about! I might suggest that `temporary point` --> `point` in a lot of these cases.

pshriwise commented on pyne/vta.py:
```diff
@@ -0,0 +1,481 @@
+            # calculate the voxel id
+            for dr in range(len(idxs)):
+                idxs[dr] = _find_voxel_idx_1d(bounds[dr], tp[dr], v[dr])
+            if -1 in idxs:
```
Why this extra check? We test if the point is in the mesh above right?

zxkjack123 replied:
The comment here should be revised. 
As the mesh has outside boundaries. A point may be exactly on the boundary. The function `_is_point_in_mesh` will give a `True` value if the point is on the boundary. 
However, we don't need to traverse a point that has a direction toward outside. This is the reason of the extra check here.

pshriwise replied:
Then will triangle on the boundary of the mesh not be counted as intersecting with the voxel? Should they be?

I'm feeling like this is a corner case that will be exposed later on...

pshriwise commented on pyne/vta.py:
```diff
@@ -0,0 +1,481 @@
+            t_maxs = _cal_max_travel_length_in_current_voxel(tp, end, v, bounds)
+            idxs, tp, t_temp = _move_to_next_voxel(idxs, tp, t_temp, t_maxs, v)
+        else:
+            # Check whether the ray intersecs the mesh. Calculate the maxinum travel
```
maxinum --> maximum

pshriwise commented on pyne/vta.py:
```diff
@@ -0,0 +1,481 @@
+        else:
+            # Check whether the ray intersecs the mesh. Calculate the maxinum travel
+            # length of the ray to the outside boundaries of the mesh. If all the
+            # length are negtive, then there is no intersection, return. Ohterwise,
```
length --> lengths

pshriwise commented on pyne/vta.py:
```diff
@@ -0,0 +1,481 @@
+    t_temp += t_min
+    return idxs, tp, t_temp
+
+def _move_to_boundary(tp, t_min, v):
```
Is this function used anywhere? It seems like there is a nice opportunity to extend it's capability and replace some code in the `_ray_voxel_traverse` function (see below).

zxkjack123 replied:
No, this function is not used anywhere because it's too simple, only one line. I will remove this function.

pshriwise commented on pyne/vta.py:
```diff
@@ -0,0 +1,481 @@
+            # length are negtive, then there is no intersection, return. Ohterwise,
+            # choose the minimum length and move the point to that position.
+            # left, right, back, front, down, up
+            t_maxs = [0.0] * 6
```
The code here is kind of what I'd expect to see in the `_move_to_boundary` function defined above.

Ideally I think there would be two calls here:

```
intersects = _check_mesh_intersection(start, end, bounds) # check if they ray intersects the mesh
if not intersects:
    break
pnt = _move_to_boundary(pnt, bounds)
```

or something like that.

zxkjack123 replied:
Actually, these two calls are both quite simple. If they are made into functions, there would be only one or two lines.

pshriwise replied:
To me, it's less about the number of lines and more about separating the functionality, which you've done really well in other places in this module.

pshriwise replied:
To put it another way, there are many functions in this module which are only a few lines, but they serve a distinct purpose and make the code easy to follow when used in other places. I think that practice could be repeated here.

zxkjack123 replied:
I used the `_move_to_boundary` here in the new commit. 
As for the intersection checking function that checks whether the line segment intersects with the mesh is not finished by a single function. This check is finished during the `while` loop. In each single the loop step, the if clause below can't tell whether there will be intersection if it's ture, but there will definitely no intersection if it's false. 
```
if any((x > 0).all() for x in dist_maxs):
```
So, this is not an independent checking function.

pshriwise commented on pyne/vta.py:
```diff
@@ -0,0 +1,481 @@
+                           bound_list[bound_idx+1] - bound_list[bound_idx])
+    return min_step
+
+def _divide_edge(start, end, step):
```
I'll recommend this gets renamed to `_divide_tri_edge` for clarity

pshriwise commented on pyne/vta.py:
```diff
@@ -0,0 +1,481 @@
+    start : numpy array
+        An array of size 3, represents the start point.
+    insert_points : list of arrays
+        List of points.
```
"List of ray end points" perhaps?

pshriwise commented on pyne/vta.py:
```diff
@@ -0,0 +1,481 @@
+
+def _create_rays_from_triangle_facet(A, B, C, min_step):
+    """
+    Create rays form triangle for scanning.
```
form --> from

pshriwise commented on pyne/vta.py:
```diff
@@ -0,0 +1,481 @@
+        An array of size 3, represents a vertex of the triangle
+    min_step: float
+        The minimum step of the mesh grid. Used to divide the shortest edge
+        of the facet triangle. Use this value could make sure that every voxel
```
This needs some rewording...

pshriwise commented on pyne/vta.py:
```diff
@@ -0,0 +1,481 @@
+
+def _facet_voxel_traverse(A, B, C, bounds):
+    """
+    Facet voxel traverse. Calculate all the voxles that intersect with the
```
Can remove the function name in the docstring here.

pshriwise replied:
This may happen in other places too. Please check it out...

pshriwise commented on pyne/vta.py:
```diff
@@ -0,0 +1,481 @@
+def _facet_voxel_traverse(A, B, C, bounds):
+    """
+    Facet voxel traverse. Calculate all the voxles that intersect with the
+    facet. Facet is represented with 3 triangle verteice.
```
vertices

pshriwise commented on pyne/vta.py:
```diff
@@ -0,0 +1,481 @@
+    return voxels
+
+if __name__ == '__main__':
+    my_list = [[1, 1], [2, 2]]
```
I'm not understanding this behavior when `vta.py` is called as main...

pshriwise commented on tests/test_vta.py:
```diff
@@ -0,0 +1,73 @@
+    y_bound = [0, 1, 2, 3, 4, 5, 6]
+    z_bound = [0, 1]
+    bounds = [x_bound, y_bound, z_bound]
+    A = np.array([3, 5.5, 0.5])
```
I'd recommend changing `A`/`B` to `ray_start`/`ray_end`. In other places A, B, C are used to represent triangle vertices.

pshriwise commented on tests/test_vta.py:
```diff
@@ -0,0 +1,73 @@
+    z_bound = [0, 1, 2]
+    bounds = [x_bound, y_bound, z_bound]
+    A = np.array([0.3, 0.4, 0.5])
+    B = np.array([2, 2, 2])
```
see comment before this one

pshriwise commented on tests/test_vta.py:
```diff
@@ -0,0 +1,73 @@
+    results = vta._ray_voxel_traverse(bounds, A, B)
+    exp_ans = set([(2, 5, 0), (2, 4, 0), (2, 3, 0), (1, 3, 0),
+               (1, 2, 0), (1, 1, 0), (0, 1, 0), (0, 0, 0)])
+#    for i in range(len(exp_ans)):
```
Why are these lines commented?

pshriwise commented on tests/test_vta.py:
```diff
@@ -0,0 +1,73 @@
+    B = np.array([2, 2, 2])
+    results = vta._ray_voxel_traverse(bounds, A, B)
+    exp_ans = set([(0, 0, 0), (0, 0, 1), (0, 1, 1), (1, 1, 1)])
+#    for i in range(len(exp_ans)):
```
Same as above, why commented? Can they be removed?

pshriwise reviewed (changes requested):
@zxkjack123 I have a lot of comments, but this was a real pleasure to review. You did a great job drawing boundaries between your functions and it made the higher-level functions easy to understand. I think its really close and it will be a solid addition to PyNE when it's done!

zxkjack123 committed:
rename several functions to be more verbose

zxkjack123 committed:
fix bug in move to next voxel, direction could be -1

zxkjack123 committed:
fix bug in move to next voxel, direction could be -1

zxkjack123 committed:
rename t_min  to dist_min, v to vec

zxkjack123 committed:
rename t_min  to dist_min, v to vec

zxkjack123 committed:
rename tp to point or updated_point

zxkjack123 committed:
rename tp to point or updated_point

zxkjack123 committed:
rename t_temp to dist_temp, t_max to dist_max

pshriwise commented on pyne/vta.py:
```diff
@@ -0,0 +1,481 @@
+        Coordinate of current point in the direction.
+    v_1d : float
+        Direction value.
+    bound : list of float
```
might be best to go with bounds rather than bound because the variable is a list of values

zxkjack123 committed:
add test for rays outside the mesh

zxkjack123 committed:
add test for rays outside the mesh

zxkjack123 committed:
fix the intersection judgement code

zxkjack123 committed:
clean up some comments

zxkjack123 committed:
clean up some comments

zxkjack123 committed:
use _move_to_boundary

zxkjack123 committed:
rename bound to bounds_1d, v_1d to vec_1d

zxkjack123 commented:
Major changes:
- Fixed a bug about the voxel index increase step (could be -1), and added a test case for it
- Renamed some names of functions and variables, made them more verbose
-

zxkjack123 committed:
add a test

zxkjack123 committed:
check ray on boundary

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,518 @@
+    cor : float
+        Coordinate of the point in the given direction.
+    vec_1d : float
+        Vector attribute in the given direction.
```
It is not clear how this parameter is used.  The docstring might want to describe what the purpose/effect of this parameter is.

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,518 @@
+    """
+    idx = -1
+    if vec_1d > 0:
+        for i in range(len(bounds_1d)-1):
```
linear search is inefficient.  A binary search should be straightforward.  Also - does numpy have a `find()` method, or similar that might be effective?

gonuke replied:
`numpy.searchsorted()`

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,518 @@
+        Direction from start to end, an unit vector.
+    """
+    vec = end - start
+    dist = math.sqrt(vec[0] * vec[0] + vec[1] * vec[1] + vec[2] * vec[2])
```
once you've decided to import `numpy` might be better to use `np.sqrt()`. In this case, it's probably equivalent, but it may remove the `math` dependency

gonuke replied:
`return np.divide(end-start,np.linalg.norm(end-start))`

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,518 @@
+    return vec
+
+def _find_next_grid_1d(cor, vec_1d, bounds_1d):
+    """
```
can't this reuse `_find_voxel_idx_1d`?

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,518 @@
+    dist = math.sqrt((end[0] - start[0]) * (end[0] - start[0]) +
+                     (end[1] - start[1]) * (end[1] - start[1]) +
+                     (end[2] - start[2]) * (end[2] - start[2]))
+    return dist
```
`dist = np.linalg.norm(end-start)`

zxkjack123 commented on pyne/vta.py:
```diff
@@ -0,0 +1,518 @@
+    vec = np.divide(vec, dist)
+    return vec
+
+def _find_next_grid_1d(cor, vec_1d, bounds_1d):
```
Similar to find_index_1D.

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,518 @@
+    """
+    n_x_grid = _find_next_grid_1d(point[0], vec[0], bounds[0])
+    n_y_grid = _find_next_grid_1d(point[1], vec[1], bounds[1])
+    n_z_grid = _find_next_grid_1d(point[2], vec[2], bounds[2])
```
```
next_grid = np.array[n_x_grid, n_y_grid, n_z_grid]
return np.divide(next_grid - point, vec)
```

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,518 @@
+    dist_temp : float
+        Updated travel length since the beginning of the ray."""
+
+    dist_min = min(i for i in dist_maxs if i > 0)
```
```suggestion
    dist_min = min(dist_maxs[dist_maxs>0])
```

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,518 @@
+                # there maybe intersects to the mesh
+                dist_min = min(i for i in dist_maxs if i > 0)
+                # Move the point to the boundary, calculate the new coordinate
+                point = _move_to_boundary(point, dist_min, vec)
```
`point = point + min_dist_bdry * dir_vec`

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,518 @@
+            else:
+               if any((x == start[i]).all() for x in bounds[i]):
+                   return True
+        return False
```
```
for idx in np.argwhere(a!=1.0):
   if start[i[0]] in bounds[i[0]]:
       return True
```

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,518 @@
+    """
+    min_step = float('inf')
+    for bound_list in bounds:
+        for bound_idx in range(len(bound_list)-1):
```
`min_step = min(min_step, min(bound_list[2:] - bound_list[:-1]))`

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,518 @@
+    y_list = np.linspace(start[1], end[1], num_segment)
+    z_list = np.linspace(start[2], end[2], num_segment)
+    for (x, y, z) in zip(x_list, y_list, z_list):
+        insert_points.append([x, y, z])
```
```suggestion
        insert_points.append(np.array([x, y, z]))
```

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,518 @@
+        source = (A, B)
+
+    insert_points = _divide_tri_edge(source[0], source[1], min_step)
+    rays = _create_rays_from_points(target, insert_points)
```
```suggestion
    rays = [ (target, point) for point in insert_points ]
```

gonuke reviewed (changes requested):
Here are some suggested changes based on better use of numpy

gonuke committed:
Update pyne/vta.py

zxkjack123 committed:
remove function _distance

zxkjack123 committed:
remove function _move to_boundary

zxkjack123 committed:
remove function _create_rays_from_points

zxkjack123 committed:
simplify the function _find_voxel_idx_1d

zxkjack123 committed:
clean up

zxkjack123 commented:
Major changes:
- Simplified the code using some functions in numpy

pshriwise commented:
Just a heads up for @zxkjack123 that I haven't forgotten about this. Just a very busy week. I will do my best to look over the updated changes within the next few days.

In the meantime can someone with write access to the repo (@gonuke?)  restart the CI builds. It looks like they failed due to a timeout of an ace dataset download.

pshriwise commented:
Oops I spoke too soon... @zxkjack123 it looks like the Python 3 build is failing due to a divide by zero error in the `vta.py` tests...

zxkjack123 committed:
fix a error in CI python3 about scalar conversion

zxkjack123 commented on pyne/vta.py:
```diff
@@ -0,0 +1,421 @@
+        Ray is not on a boundary surface
+    """
+    if 1.0 in vec:
+        for idx in np.argwhere(vec!=1.0):
```
```suggestion
        for idx in np.argwhere(vec!=1.0).flatten():
```

zxkjack123 committed:
add function for triangle-box intersection check

zxkjack123 commented:
Major changes:
- Add a function for triangle-box intersection check based on separate axis theorem
- Add a test function for this triangle-box intersection check

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,514 @@
+    # a, axis for test in bullet 3. aij = ei x fj, i.e., a00 = e0 x f0
+
+    # define bais axis
+    e = np.zeros(shape=(3, 3))
```
`e = np.eye(3,3)`

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,514 @@
+    e[2] = np.array([0, 0, 1])
+    # Translate triangle as conceptually moving AABB to origin
+    v = np.zeros(shape=(3, 3))
+    v[0] = triangle[0] - box_center
```
??? `v = triangle - box_center`  ??

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,514 @@
+    for i in range(0, 3):
+        for j in range(0, 3):
+            a = np.cross(e[i], f[j])
+            p0 = np.dot(v[0], a)
```
?? `p = np.mult(v,a)`

gonuke commented on pyne/vta.py:
```diff
@@ -0,0 +1,514 @@
+    ## region Test the three axes corresponding to the face normals of AABB b (category 1)
+    # Exit if...
+    # x direction
+    if max(v[:, 0]) < -box_extents[0] or min(v[:, 0]) > box_extents[0]:
```
loop over directions (0,1,2)

zxkjack123 committed:
vectorize some code in tri-box intersection check

zxkjack123 commented:
Update:
- Simplify the function tri-box intersection check via vectorizing some variables

moatazharb commented on pyne/vta.py:
```diff
@@ -0,0 +1,500 @@
+        return True
+    else:
+        return False
+
```
Why not check coordinates for being outside the boundaries in their respective directions?. All "AND" statements will be executed unless a "False" is found which means that all statements will be checked if they result in True. I suggest switching to "NOT OR" statements. This way if at least one coordinate is outside the mesh then there will be no need to continue checking the rest of the directions. 
Note: I might not have succeeded in explaining what I mean but you can check "short circuit" of logical operators.
```suggestion
    if not (bounds[0][0] <= point[0] <= bounds[0][-1]) or \
      not (bounds[1][0] <= point[1] <= bounds[1][-1]) or \
      not (bounds[2][0] <= point[2] <= bounds[2][-1]):
        return False
    else:
        return True
```

moatazharb commented on pyne/vta.py:
```diff
@@ -0,0 +1,500 @@
+    ----------
+    """
+    idx = -1
+    bounds_1d = np.array(bounds_1d)
```
What is the format of bounds_1d?. Is it from the nested list you mentioned in "_is_point_in_mesh" function?. If so then you may consider making the nested list into an array with each row containing the boundaries along one dimension;
bounds = np.array([[x_min, ..., x_max],
                                [y_min, ..., y_max],
                                [z_min, ..., z_max]]) 
This way bounds_1d = bounds[0] (or [1] or [2] according yo which direction is considered) and it is a numpy array so you wouldn't need to cast bounds_1d as an array here.

zxkjack123 replied:
Applied this change. Numpy array is better than the nested list in this case.

moatazharb commented on pyne/vta.py:
```diff
@@ -0,0 +1,500 @@
+    else:
+        return idx - 1
+
+def _calc_vec_dir(start, end):
```
One line!, probably no need for a function since it's too simple. Even if it's used many times in many functions, it is still one line and no need to be included in a separate function. Up to you :)

zxkjack123 replied:
Though it's only one line, but I think the function name `calc_vec_dir` are easier to understand than the code. So, I will keep this function.

moatazharb commented on pyne/vta.py:
```diff
@@ -0,0 +1,500 @@
+    """
+    voxel_idx = _find_voxel_idx_1d(bounds_1d, cor, vec_1d)
+    if voxel_idx == -1:
+        return vec_1d * float('inf')
```
Why not np.inf?. [https://stackoverflow.com/questions/42315541/difference-between-np-inf-and-floatinf](https://stackoverflow.com/questions/42315541/difference-between-np-inf-and-floatinf)

zxkjack123 committed:
Change coordinate check for box

zxkjack123 committed:
change bounds from list to numpy array

zxkjack123 committed:
change float('inf') to np.inf

moatazharb commented on pyne/vta.py:
```diff
@@ -0,0 +1,500 @@
+    n_y_grid = _find_next_grid_1d(point[1], vec[1], bounds[1])
+    n_z_grid = _find_next_grid_1d(point[2], vec[2], bounds[2])
+    next_grid = np.array([n_x_grid, n_y_grid, n_z_grid])
+    return np.divide(next_grid - point, vec)
```
Why divide by the unit vector "vec"?. If I understand correctly, you calculated the distance from the point to the next grid boundaries and you need to get the components that correspond to a vector from the point to the boundaries along "vec", so I guess you need the projection of "next_grid - point" on "vec!

Also create a variable and add a comment above to describe what you're calculating then return it
```suggestion
    # Calculate .......
    dist_maxs = np.divide(next_grid - point, vec)
    return dist_maxs
```

moatazharb commented on pyne/vta.py:
```diff
@@ -0,0 +1,500 @@
+    n_y_grid = _find_next_grid_1d(point[1], vec[1], bounds[1])
+    n_z_grid = _find_next_grid_1d(point[2], vec[2], bounds[2])
+    next_grid = np.array([n_x_grid, n_y_grid, n_z_grid])
+    return np.divide(next_grid - point, vec)
```
Can you add a comment above the np.divide calculation to describe it and also create a variable and return it?.
```suggestion
    # Divide by direction cosines to calculate the distance along vec
    dist_maxs = np.divide(next_grid - point, vec)
    return dist_maxs
```

moatazharb replied:
I think also that the function should return the minimum of all three distances as this would be the maximum travel distance in the voxel along vec before reaching a boundary.

moatazharb commented on pyne/vta.py:
```diff
@@ -0,0 +1,500 @@
+        update_idx[update_dir] = 1
+    elif vec[update_dir] < 0:
+        update_idx[update_dir] = -1
+    idxs = [x + y for x, y in zip(idxs, update_idx)] 
```
Why not just:
```suggestion
    # vec[update_dir] will not be 0
    if vec[update_dir] > 0:
        idxs[update_dir] += 1
    elif vec[update_dir] < 0:
        idxs[update_dir] -= 1 
```

Commit:
Update pyne/vta.py

Create a variable before returning the max travel length in a voxel

Co-Authored-By: zxkjack123 <zxkjack123@163.com>

Commit:
Update pyne/vta.py

Simply the idx change in traverse.

Co-Authored-By: zxkjack123 <zxkjack123@163.com>

zxkjack123 committed:
clean up

moatazharb commented on pyne/vta.py:
```diff
@@ -0,0 +1,500 @@
+    """
+    # Initialization
+    # Find the voxel that start point in
+    point = np.copy(start)
```
maybe
```suggestion
    point = start[:]
```

moatazharb commented on pyne/vta.py:
```diff
@@ -0,0 +1,500 @@
+    vec = _calc_vec_dir(start, end)
+    # if the ray on a boundary, then return empty set
+    if _is_ray_on_boundary(start, vec, bounds):
+        return set()
```
for consistency, why not just return "voxels" since this is the first check after creating the set it'll be empty as you needed it to be if the if statement is True.
```suggestion
        return voxels
```

moatazharb commented on pyne/vta.py:
```diff
@@ -0,0 +1,500 @@
+        return set()
+    dist_end = np.linalg.norm(end-start)
+    dist_temp = 0.0
+    while dist_temp < dist_end:
```
There is a possible problem here; dist_temp not being exactly = dist_end in which case the while loop will not terminate. 
A better way to do this is to (add) check the the index of the voxel of the current point against the end point.
Another way is to use a comparison within some rounding limit like numpy.isclose.

zxkjack123 replied:
The `dist_temp` here is the distance  from the start point of the ray to the intersection point of the ray with the mesh boundary. The `dist_temp` could be `inf` if the ray goes out of the mesh. So, the `dist_temp` will always be greater than the `dist_end` (a definite value) in several iterations. Therefore, the loop will terminate.

moatazharb commented on pyne/vta.py:
```diff
@@ -0,0 +1,500 @@
+            # choose the minimum length and move the point to that position.
+            # left, right, back, front, down, up
+            dist_maxs = [0.0] * 6
+            for dr in [0, 1, 2]:
```
For consistency, why not
```suggestion
            for dr in range(len(idxs):
```

moatazharb commented on pyne/vta.py:
```diff
@@ -0,0 +1,500 @@
+            if any((x > 0).all() for x in dist_maxs):
+                # current point outside the mesh
+                # there maybe intersects to the mesh
+                dist_min = min(i for i in dist_maxs if i > 0)
```
```suggestion
                dist_min = min(dist_maxs[dist_maxs > 0])
```

moatazharb commented on pyne/vta.py:
```diff
@@ -0,0 +1,500 @@
+    # f, edge vector of the triangle.  f0=v1-v0, f1=v2-v1, f2=v0-v2
+    # a, axis for test in bullet 3. aij = ei x fj, i.e., a00 = e0 x f0
+
+    # define bais axis
```
typo, "basis"

moatazharb commented on pyne/vta.py:
```diff
@@ -0,0 +1,500 @@
+    # a, axis for test in bullet 3. aij = ei x fj, i.e., a00 = e0 x f0
+
+    # define bais axis
+    e = np.eye(3, 3)
```
Maybe call it unit_vec or axes_unit_vecs. Similarly for "v", "f", and "a". It makes it easier to follow when the name of the variable says something about it.

zxkjack123 replied:
Well, the codes that use the separate axis theorem (SAT) are converted from the other people's codes which are based on several papers related to SAT. In the se papers/codes, the `e`, `a`, `f` and `v` are used.
In my opinion, keep these variables similar to the papers introducing SAT and the codes that apply SAT to the `box-triangle` check would be a better choice.

moatazharb commented on pyne/vta.py:
```diff
@@ -0,0 +1,500 @@
+    """
+
+    # variables names used in the code
+    # e, unit vector of axis. e0=(1, 0, 0), e1=(0, 1, 0), e2=(0, 0, 1).
```
wouldn't it be better to call it ex, ey, ez since it's aligned witht the axes?

zxkjack123 replied:
I changed them to `e[0], e[1], e[2]`.

moatazharb commented on pyne/vta.py:
```diff
@@ -0,0 +1,500 @@
+    f[2] = triangle[0] - triangle[2]
+
+    ## region Test axes a00..a22 (category 3)
+    for i in range(0, 3):
```
just range(3) should be fine

moatazharb commented on pyne/vta.py:
```diff
@@ -0,0 +1,500 @@
+    ## region Test axes a00..a22 (category 3)
+    for i in range(0, 3):
+        for j in range(0, 3):
+            a = np.cross(e[i], f[j])
```
Add description before each of these products. like: # Calculating the normal to edge-axis plane, etc.

moatazharb commented on pyne/vta.py:
```diff
@@ -0,0 +1,500 @@
+            a = np.cross(e[i], f[j])
+            p = np.dot(v, a)
+            r = np.dot(box_extents, np.absolute(a))
+            if min(p) > r or max(p) < -r:
```
Add description of what this condition means!.

Commit:
Update pyne/vta.py

Co-Authored-By: zxkjack123 <zxkjack123@163.com>

Commit:
Update pyne/vta.py

Co-Authored-By: zxkjack123 <zxkjack123@163.com>

Commit:
Update pyne/vta.py

Co-Authored-By: zxkjack123 <zxkjack123@163.com>

zxkjack123 committed:
update pyne/vta.py according to Moataz's comments

zxkjack123 committed:
rebase and update

zxkjack123 committed:
add comment content to trigger CI

zxkjack123 committed:
trigger CI

zxkjack123 committed:
trigger new CI

zxkjack123 committed:
CI 4th

zxkjack123 committed:
clean up

zxkjack123 commented:
Close this PR. Because the margin for speed up particle birth is too small. I tested with a R2S photon transport using FNG model, the CPU time of particle birth takes about only 1.5% when #1118 is applied. It's worthless to do source sampling with bounding box.
